File: Assets/WebRtcNetwork/Byn.Net/LocalNetwork.cs
//using Byn.Common;
//using System;
//using System.Collections.Generic;
//using System.Linq;

//namespace Byn.Net
//{
//	public class LocalNetwork : ServerConnection, INetwork, IDisposable
//	{
//		private class WeakRef<T> where T : class
//		{
//			private WeakReference mRef;

//			public WeakRef(T val)
//			{
//				this.mRef = new WeakReference(val);
//			}

//			public T Get()
//			{
//				if (this.mRef.IsAlive)
//				{
//					return this.mRef.Target as T;
//				}
//				return default(T);
//			}
//		}

//		private static int sNextId = 1;

//		private static Dictionary<string, LocalNetwork.WeakRef<LocalNetwork>> mServers = new Dictionary<string, LocalNetwork.WeakRef<LocalNetwork>>();

//		private int mId;

//		private ConnectionId mNextNetworkId = new ConnectionId
//		{
//			id = 1
//		};

//		private string mServerAddress;

//		private Queue<NetworkEvent> mEvents = new Queue<NetworkEvent>();

//		private Dictionary<ConnectionId, LocalNetwork.WeakRef<LocalNetwork>> mConnectionNetwork = new Dictionary<ConnectionId, LocalNetwork.WeakRef<LocalNetwork>>();

//		private bool disposedValue;

//		public IList<ConnectionId> Connections
//		{
//			get
//			{
//				return this.mConnectionNetwork.Keys.ToList<ConnectionId>();
//			}
//		}

//		public bool IsServer
//		{
//			get
//			{
//				return this.mServerAddress != null;
//			}
//		}

//		public LocalNetwork()
//		{
//			this.mId = LocalNetwork.sNextId;
//			LocalNetwork.sNextId++;
//		}

//        public void LeaveRoom()
//        {
//            if (this.IsServer)
//            {
//                this.Enqueue(NetEventType.ServerClosed, ConnectionId.INVALID, this.mServerAddress);
//                LocalNetwork.mServers.Remove(this.mServerAddress);
//                this.mServerAddress = null;
//            }
//        }

//        public void CreateRoom(string serverAddress = null)
//        {
//            if (serverAddress == null)
//            {
//                serverAddress = string.Concat(this.mId);
//            }
//            if (LocalNetwork.mServers.ContainsKey(serverAddress) && LocalNetwork.mServers[serverAddress].Get() != null)
//            {
//                this.Enqueue(NetEventType.ServerConnectionFailed, ConnectionId.INVALID, serverAddress);
//                return;
//            }
//            LocalNetwork.mServers[serverAddress] = new LocalNetwork.WeakRef<LocalNetwork>(this);
//            this.mServerAddress = serverAddress;
//            this.Enqueue(NetEventType.ServerInitialized, ConnectionId.INVALID, serverAddress);
//        }

//		public void ConnectToServer()
//		{
//		}

//		public void Disconnect()
//		{
			
//		}

//		public ConnectionId ConnectToRoom(string address)
//		{
//			ConnectionId connectionId = this.NextConnectionId();
//			bool sucessful = false;
//			if (LocalNetwork.mServers.ContainsKey(address))
//			{
//				LocalNetwork server = LocalNetwork.mServers[address].Get();
//				if (server != null)
//				{
//					server.ConnectClient(this);
//					this.mConnectionNetwork[connectionId] = LocalNetwork.mServers[address];
//					this.Enqueue(NetEventType.NewConnection, connectionId, null);
//					sucessful = true;
//				}
//			}
//			if (!sucessful)
//			{
//				this.Enqueue(NetEventType.ConnectionFailed, connectionId, "Couldn't connect to the given server with id " + address);
//			}
//			return connectionId;
//		}

//		public void Shutdown()
//		{
//			foreach (ConnectionId v in this.mConnectionNetwork.Keys.ToList<ConnectionId>())
//			{
//				this.Disconnect(v);
//			}
//			this.mConnectionNetwork.Clear();
//			this.Disconnect();
//		}

//		public void SendData(ConnectionId userId, byte[] data, int offset, int length, bool reliable)
//		{
//			LocalNetwork.WeakRef<LocalNetwork> netref;
//			if (this.mConnectionNetwork.TryGetValue(userId, out netref))
//			{
//				this.mConnectionNetwork[userId].Get().ReceiveData(this, data, offset, length, reliable);
//			}
//		}

//		public void Update()
//		{
//			this.CleanupWreakReferences();
//		}

//		public bool Dequeue(out NetworkEvent evt)
//		{
//			if (this.mEvents.Count == 0)
//			{
//				evt = default(NetworkEvent);
//				return false;
//			}
//			evt = this.mEvents.Dequeue();
//			return true;
//		}

//		public bool Peek(out NetworkEvent evt)
//		{
//			if (this.mEvents.Count == 0)
//			{
//				evt = default(NetworkEvent);
//				return false;
//			}
//			evt = this.mEvents.Peek();
//			return true;
//		}

//		public void Flush()
//		{
//		}

//		public void Disconnect(ConnectionId id)
//		{
//			if (this.mConnectionNetwork.ContainsKey(id))
//			{
//				LocalNetwork other = this.mConnectionNetwork[id].Get();
//				if (other != null)
//				{
//					other.InternalDisconnect(this);
//					this.InternalDisconnect(id);
//					return;
//				}
//				this.CleanupWreakReferences();
//			}
//		}

//		private ConnectionId FindConnectionId(LocalNetwork network)
//		{
//			foreach (KeyValuePair<ConnectionId, LocalNetwork.WeakRef<LocalNetwork>> kvp in this.mConnectionNetwork)
//			{
//				if (kvp.Value.Get() == network)
//				{
//					return kvp.Key;
//				}
//			}
//			return ConnectionId.INVALID;
//		}

//		private ConnectionId NextConnectionId()
//		{
//			ConnectionId arg_17_0 = this.mNextNetworkId;
//			this.mNextNetworkId.id = (short)(this.mNextNetworkId.id + 1);
//			return arg_17_0;
//		}

//		private void ConnectClient(LocalNetwork client)
//		{
//			if (!this.IsServer)
//			{
//				throw new InvalidOperationException();
//			}
//			ConnectionId nextId = this.NextConnectionId();
//			this.mConnectionNetwork[nextId] = new LocalNetwork.WeakRef<LocalNetwork>(client);
//			this.Enqueue(NetEventType.NewConnection, nextId, null);
//		}

//		private void Enqueue(NetEventType type, ConnectionId id, object data)
//		{
//			NetworkEvent ev = new NetworkEvent(type, id, data);
//			this.mEvents.Enqueue(ev);
//		}

//		private void ReceiveData(LocalNetwork network, byte[] data, int offset, int length, bool reliable)
//		{
//			ConnectionId userId = this.FindConnectionId(network);
//			ByteArrayBuffer buffer = ByteArrayBuffer.Get(length, false);
//			buffer.CopyFrom(data, offset, length);
//			if (buffer.PositionReadRelative != 0)
//			{
//				SLog.LE("Received an invalid message. PositionReadRelative must be 0 but is " + buffer.PositionReadRelative, new string[0]);
//			}
//			if (buffer.PositionWriteRelative == 0)
//			{
//				SLog.LE("Received an invalid message. Message empty.", new string[0]);
//			}
//			NetEventType type = NetEventType.UnreliableMessageReceived;
//			if (reliable)
//			{
//				type = NetEventType.ReliableMessageReceived;
//			}
//			this.Enqueue(type, userId, buffer);
//		}

//		private void InternalDisconnect(ConnectionId id)
//		{
//			if (this.mConnectionNetwork.ContainsKey(id))
//			{
//				this.Enqueue(NetEventType.Disconnected, id, null);
//				this.mConnectionNetwork.Remove(id);
//			}
//		}

//		private void InternalDisconnect(LocalNetwork ln)
//		{
//			this.InternalDisconnect(this.FindConnectionId(ln));
//		}

//		private void CleanupWreakReferences()
//		{
//			foreach (ConnectionId kvp in this.mConnectionNetwork.Keys.ToList<ConnectionId>())
//			{
//				if (this.mConnectionNetwork[kvp].Get() == null)
//				{
//					this.InternalDisconnect(kvp);
//				}
//			}
//		}

//		protected virtual void Dispose(bool disposing)
//		{
//			if (!this.disposedValue)
//			{
//				if (disposing)
//				{
//					this.Shutdown();
//				}
//				this.disposedValue = true;
//			}
//		}

//		public void Dispose()
//		{
//			this.Dispose(true);
//		}
//	}
//}

File: Assets/WebGLTemplates/Main/WebRTC/LocalNetwork.js
var LocalNetwork = function() {
    function e() {
        this.mNextNetworkId = new ConnectionId(1);
        this.mServerAddress = null;
        this.mEvents = new Queue;
        this.mConnectionNetwork = {};
        this.mIsDisposed = false;
        this.mId = e.sNextId;
        e.sNextId++
    }
    Object.defineProperty(e.prototype, "IsServer", {
        get: function() {
            return this.mServerAddress != null
        },
        enumerable: true,
        configurable: true
    });
    e.prototype.CreateRoom = function(t) {
        if (t === void 0) {
            t = null
        }
        if (t == null) t = "" + this.mId;
        if (t in e.mServers) {
            this.Enqueue(NetEventType.ServerInitFailed, ConnectionId.INVALID, t);
            return
        }
        e.mServers[t] = this;
        this.mServerAddress = t;
        this.Enqueue(NetEventType.ServerInitialized, ConnectionId.INVALID, t)
    };
    e.prototype.LeaveRoom = function() {
        if (this.IsServer) {
            this.Enqueue(NetEventType.ServerClosed, ConnectionId.INVALID, this.mServerAddress);
            delete e.mServers[this.mServerAddress];
            this.mServerAddress = null
        }
    };
    e.prototype.ConnectToRoom = function(t) {
        var n = this.NextConnectionId();
        var i = false;
        if (t in e.mServers) {
            var o = e.mServers[t];
            if (o != null) {
                o.ConnectClient(this);
                this.mConnectionNetwork[n.id] = e.mServers[t];
                this.Enqueue(NetEventType.NewConnection, n, null);
                i = true
            }
        }
        if (i == false) {
            this.Enqueue(NetEventType.ConnectionFailed, n, "Couldn't connect to the given server with id " + t)
        }
        return n
    };
    e.prototype.Shutdown = function() {
        for (var e in this.mConnectionNetwork) {
            this.DisconnectPeerFromServer(new ConnectionId((+e)))
        }
        this.LeaveRoom()
    };
    e.prototype.Dispose = function() {
        if (this.mIsDisposed == false) {
            this.Shutdown()
        }
    };
    e.prototype.SendData = function(e, t, n) {
        if (e.id in this.mConnectionNetwork) {
            var i = this.mConnectionNetwork[e.id];
            i.ReceiveData(this, t, n)
        }
    };
    e.prototype.Update = function() {
        this.CleanupWreakReferences()
    };
    e.prototype.Dequeue = function() {
        return this.mEvents.Dequeue()
    };
    e.prototype.Peek = function() {
        return this.mEvents.Peek()
    };
    e.prototype.Flush = function() {};
    e.prototype.DisconnectPeerFromServer = function(e) {
        if (e.id in this.mConnectionNetwork) {
            var t = this.mConnectionNetwork[e.id];
            if (t != null) {
                t.InternalDisconnectNetwork(this);
                this.InternalDisconnect(e)
            } else {
                this.CleanupWreakReferences()
            }
        }
    };
    e.prototype.FindConnectionId = function(e) {
        for (var t in this.mConnectionNetwork) {
            var n = this.mConnectionNetwork[t];
            if (n != null) {
                return new ConnectionId((+t))
            }
        }
        return ConnectionId.INVALID
    };
    e.prototype.NextConnectionId = function() {
        var e = this.mNextNetworkId;
        this.mNextNetworkId = new ConnectionId(e.id + 1);
        return e
    };
    e.prototype.ConnectClient = function(e) {
        var t = this.NextConnectionId();
        this.mConnectionNetwork[t.id] = e;
        this.Enqueue(NetEventType.NewConnection, t, null)
    };
    e.prototype.Enqueue = function(e, t, n) {
        var i = new NetworkEvent(e, t, n);
        this.mEvents.Enqueue(i)
    };
    e.prototype.ReceiveData = function(e, t, n) {
        var i = this.FindConnectionId(e);
        var o = new Uint8Array(t.length);
        for (var r = 0; r < o.length; r++) {
            o[r] = t[r]
        }
        var a = NetEventType.UnreliableMessageReceived;
        if (n) a = NetEventType.ReliableMessageReceived;
        this.Enqueue(a, i, o)
    };
    e.prototype.InternalDisconnect = function(e) {
        if (e.id in this.mConnectionNetwork) {
            this.Enqueue(NetEventType.Disconnected, e, null);
            delete this.mConnectionNetwork[e.id]
        }
    };
    e.prototype.InternalDisconnectNetwork = function(e) {
        this.InternalDisconnect(this.FindConnectionId(e))
    };
    e.prototype.CleanupWreakReferences = function() {};
    e.sNextId = 1;
    e.mServers = {};
    return e
}();
